base/utf16.jl

```diff
@@ -71,25 +152,346 @@ function encode16(s::AbstractString)
+                @inbounds ch = ((ch & 0x1f) << 6) | (iStr[pos += 1] & 0x3f)
+            elseif ch < 0xf0
+                @inbounds ch = ((ch & 0xf) << 12) | (UInt32(iStr[pos+1] & 0x3f) << 6) | (iStr[pos+2] & 0x3f)
+                pos += 2
+            else
+                @inbounds ch = ((ch & 0x7) << 18) | (UInt32(iStr[pos+1] & 0x3f) << 12) | (UInt32(iStr[pos+2] & 0x3f) << 6) | (iStr[pos+3] & 0x3f)
+                pos += 3
+                @inbounds buf[out += 1] = 0xd7c0 + (ch >>> 10)
+                ch = 0xdc00 + (ch & 0x3ff)
+            end
+        end
+        @inbounds buf[out += 1] = ch
+    end
+    UTF16String(buf)
+end
+
+#=
+@doc """
+@brief      Converts a UTF-16 encoded string to UTF-8 encoding
+
+@param[in]  iStr::UTF16String
+
+@return     ::UTF8String
+@throws     UnicodeError
+""" ->
+=#
+function utf16_to_utf8(iStr::Array{UInt16,1})
+    # handle zero length string quickly
+    sizeof(iStr) <= 2 && return UTF8String("")
+    # get number of bytes to allocate
+    len::UInt, cnt2::UInt, cnt3::UInt, cnt4::UInt, flags::UInt = utf16_validate(iStr)
+    len += cnt2 + cnt3*2 + cnt4*3
+    buf = Array{UInt8,1}(len)
+    if flags == 0
+        # If copy! is fixed to handle narrowing operations efficiently, change this to use copy!
+        # 7-bit ASCII characters only! Do optimized conversion
+        for pos=1:len
+            @inbounds buf[pos] = iStr[pos]
+        end
+    else
+        out::UInt = 0
+        pos::UInt = 0
+        while out < len
+            @inbounds ch::UInt32 = iStr[pos += 1]
+            if ch > 0x7f
+                if ch < 0x800
+                    @inbounds buf[out += 1] = 0xc0 | (ch >>> 6)
+                elseif is_surrogate_char(ch)
+                    @inbounds ch = 0x10000 + (((ch & 0x3ff) << 10) | (iStr[pos += 1] & 0x3ff))
+                    @inbounds buf[out += 1] = 0xf0 | (ch >>> 18)
+                    @inbounds buf[out += 1] = 0x80 | ((ch >>> 12) & 0x3f)
+                    @inbounds buf[out += 1] = 0x80 | ((ch >>> 6) & 0x3f)
+                else
+                    @inbounds buf[out += 1] = 0xe0 | ((ch >>> 12) & 0x3f)
+                    @inbounds buf[out += 1] = 0x80 | ((ch >>> 6) & 0x3f)
+                end
+                ch = 0x80 | (ch & 0x3f)
+            end
+            @inbounds buf[out += 1] = ch
+        end
+    end
+    UTF8String(buf)
+end
+
 utf16(x) = convert(UTF16String, x)
 convert(::Type{UTF16String}, s::UTF16String) = s
+convert(::Type{UTF16String}, s::UTF8String) = utf8_to_utf16(s.data)
+convert(::Type{UTF16String}, s::ASCIIString) = ascii_to_utf16(s.data)
 convert(::Type{UTF16String}, s::AbstractString) = encode16(s)
 convert(::Type{Array{UInt16,1}}, s::UTF16String) = s.data
 convert(::Type{Array{UInt16}}, s::UTF16String) = s.data
-
-# TODO: optimize this
-convert(::Type{UTF8String}, s::UTF16String) =
-    sprint(length(s.data)-1, io->for c in s; write(io,c::Char); end)
+convert(::Type{UTF8String}, s::UTF16String) = utf16_to_utf8(s.data)
```

Comment:
Why have a separate function here, the body of `utf16_to_utf8` could just be the definition for the conversion.  Calling out to a helper function is fine if you reuse it more than once, but I don't think you do in this PR.